src/core/storage/fileio/s3_filesys.hpp

```diff
@@ -0,0 +1,384 @@
+  kDirectory
+};
+
+/*! \brief use to store file information */
+struct FileInfo {
+  /*! \brief full path to the file */
+  s3url path;
+  /*! \brief the size of the file */
+  size_t size;
+  /*! \brief the type of the file */
+  FileType type;
+  /*! \brief default constructor */
+  FileInfo() : size(0), type(kFile) {}
+};
+
+struct ScopedAwsInitAPI {
+  ScopedAwsInitAPI(const Aws::SDKOptions &options) : options_(options) {
+    Aws::InitAPI(options_);
+  }
+  ~ScopedAwsInitAPI() { Aws::ShutdownAPI(options_); }
+  Aws::SDKOptions options_;
+};
+
+const ScopedAwsInitAPI &turi_global_AWS_SDK_setup(
+    const Aws::SDKOptions &options = Aws::SDKOptions());
+
+class Stream {
+ public:
+  /*!
+   * \brief reads data from a stream
+   * \param ptr pointer to a memory buffer
+   * \param size block size
+   * \return the size of data read
+   */
+  virtual size_t Read(void *ptr, size_t size) = 0;
+  /*!
+   * \brief writes data to a stream
+   * \param ptr pointer to a memory buffer
+   * \param size block size
+   */
+  virtual void Write(const void *ptr, size_t size) = 0;
+
+  /*!
+   * \brief closes the stream
+   */
+  virtual void Close() = 0;
+
+  /*! \brief virtual destructor */
+  virtual ~Stream(void) {}
+  /*!
+   * \brief generic factory function
+   *  create an stream, the stream will close the underlying files upon deletion
+   *
+   * \param uri the uri of the input currently we support
+   *            hdfs://, s3://, and file:// by default file:// will be used
+   * \param flag can be "w", "r", "a"
+   * \param allow_null whether NULL can be returned, or directly report error
+   * \return the created stream, can be NULL when allow_null == true and file do
+   * not exist
+   */
+  static Stream *Create(const char *uri, const char *const flag,
+                        bool allow_null = false);
+};
+
+/*! \brief interface of i/o stream that support seek */
+class SeekStream : public Stream {
+ public:
+  // virtual destructor
+  virtual ~SeekStream(void) {}
+  /*! \brief seek to certain position of the file */
+  virtual void Seek(size_t pos) = 0;
+
+  /*! \brief tell the position of the stream */
+  virtual size_t Tell(void) = 0;
+
+  /*! tell the physical size of the stream */
+  virtual size_t FileSize(void) const = 0;
+
+  /*! \brief Returns true if at end of stream */
+  virtual bool AtEnd(void) const = 0;
+  /*!
+   * \brief generic factory function
+   *  create an SeekStream for read only,
+   *  the stream will close the underlying files upon deletion
+   *  error will be reported and the system will exit when create failed
+   * \param uri the uri of the input currently we support
+   *            hdfs://, s3://, and file:// by default file:// will be used
+   * \param allow_null whether NULL can be returned, or directly report error
+   * \return the created stream, can be NULL when allow_null == true and file do
+   * not exist
+   */
+  static SeekStream *CreateForRead(const char *uri, bool allow_null = false);
```

Comment:
Avoid using raw pointers in any situation where they are not isolated to a single scope.

Reply:
Yeah, inherits from the previous implementation. Will change it to smart pointer.